waiter/src/waiter/token.clj

```diff
@@ -235,24 +246,34 @@
                                                             (remove #(contains? new-service-description-template %1)) seq)
                                    :service-description new-service-description-template}))))
-              (when (and run-as-user (not= "*" run-as-user))
-                (when-not (can-run-as? authenticated-user run-as-user)
-                  (throw (ex-info "Cannot run as user" {:authenticated-user authenticated-user :run-as-user run-as-user}))))
-              (let [existing-service-description-owner (get existing-token-metadata "owner")]
-                (if-not (str/blank? existing-service-description-owner)
-                  (when-not (can-run-as? authenticated-user existing-service-description-owner)
-                    (throw (ex-info "Cannot change owner of token"
-                                    {:existing-owner existing-service-description-owner
-                                     :new-user owner})))
-                  (when-not (can-run-as? authenticated-user owner)
-                    (throw (ex-info "Cannot create token as user" {:authenticated-user authenticated-user :owner owner})))))
+              (case (get request-params "update-mode")
+                "admin"
+                (when-not (authz/administer-token? entitlement-manager authenticated-user token new-token-metadata)
+                  (throw (ex-info "Cannot administer token" {:status 403, :token-metadata new-token-metadata, :user authenticated-user})))
+
+                nil
+                (do
+                  (when (and run-as-user (not= "*" run-as-user))
+                    (when-not (authz/run-as? entitlement-manager authenticated-user run-as-user)
+                      (throw (ex-info "Cannot run as user" {:authenticated-user authenticated-user, :run-as-user run-as-user, :status 403}))))
+                  (let [existing-service-description-owner (get existing-token-metadata "owner")]
+                    (if-not (str/blank? existing-service-description-owner)
+                      (when-not (authz/manage-token? entitlement-manager authenticated-user token existing-token-metadata)
+                        (throw (ex-info "Cannot change owner of token" {:existing-owner existing-service-description-owner, :new-user owner, :status 403})))
+                      (when-not (authz/run-as? entitlement-manager authenticated-user owner)
+                        (throw (ex-info "Cannot create token as user" {:authenticated-user authenticated-user, :owner owner, :status 403}))))))
+
+                (throw (ex-info "Invalid update-mode" {:mode (get request-params "update-mode"), :status 403})))
+
               ; Store the token
-              (store-service-description-for-token synchronize-fn kv-store token new-service-description-template {"owner" owner})
+              (let [new-token-metadata (merge {"last-update-time" (.getMillis ^DateTime (clock))
```

Comment:
If `admin` mode, you need to accept `last-update-time` data and use that for the token so that you can perform a sync.